File: apps/api/app/db/models/tutoring.py
import uuid
from datetime import datetime
from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text, Boolean
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from app.db.base import Base, TenantMixin


class TutoringSession(Base, TenantMixin):
    __tablename__ = "tutoring_sessions"

    tutor_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    subject_id = Column(UUID(as_uuid=True), ForeignKey("subjects.id"), nullable=True)
    session_date = Column(DateTime, nullable=False)
    duration_minutes = Column(Integer, nullable=False, default=60)
    rate_cents = Column(Integer, nullable=False, default=0)  # per-session rate
    status = Column(String(20), default="completed")  # scheduled|completed|cancelled|no_show
    notes = Column(Text, nullable=True)  # private tutor notes
    homework_given = Column(Text, nullable=True)  # visible to student/parent
    package_id = Column(UUID(as_uuid=True), ForeignKey("tutoring_packages.id"), nullable=True)
    invoiced = Column(Boolean, default=False, nullable=False)


class TutoringPackage(Base, TenantMixin):
    __tablename__ = "tutoring_packages"

    tutor_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    name = Column(String(255), nullable=False)  # e.g. "10 sessions maths"
    sessions_total = Column(Integer, nullable=False)
    sessions_used = Column(Integer, default=0, nullable=False)
    price_cents = Column(Integer, nullable=False)  # total package price
    status = Column(String(20), default="active")  # active|completed|cancelled
    paid = Column(Boolean, default=False, nullable=False)
    notes = Column(Text, nullable=True)


class TutoringInvoice(Base, TenantMixin):
    __tablename__ = "tutoring_invoices"

    tutor_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    student_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    invoice_number = Column(String(50), nullable=False)
    period_label = Column(String(50), nullable=True)  # e.g. "March 2026"
    line_items = Column(JSONB, default=list)  # [{desc, qty, unit_price_cents, total_cents}]
    total_cents = Column(Integer, nullable=False, default=0)
    paid = Column(Boolean, default=False, nullable=False)
    paid_at = Column(DateTime, nullable=True)
    notes = Column(Text, nullable=True)

File: apps/web/src/api/tutoring.ts
import client from './client';

export interface TutoringSessionData {
  id: string;
  tutor_id: string;
  student_id: string;
  subject_id: string | null;
  session_date: string;
  duration_minutes: number;
  rate_cents: number;
  status: 'scheduled' | 'completed' | 'cancelled' | 'no_show';
  notes: string | null;
  homework_given: string | null;
  package_id: string | null;
  invoiced: boolean;
  created_at: string;
}

export interface TutoringPackageData {
  id: string;
  tutor_id: string;
  student_id: string;
  name: string;
  sessions_total: number;
  sessions_used: number;
  price_cents: number;
  status: 'active' | 'completed' | 'cancelled';
  paid: boolean;
  notes: string | null;
}

export interface TutoringInvoiceData {
  id: string;
  invoice_number: string;
  student_id: string;
  period_label: string | null;
  line_items: Array<{ description: string; total_cents: number }>;
  total_cents: number;
  paid: boolean;
  paid_at: string | null;
  created_at: string;
}

export interface MyStudent {
  student_id: string;
  session_count: number;
  last_session: string | null;
}

export function getMyStudents() { return client.get<MyStudent[]>('/v1/tutoring/my-students'); }
export function listSessions(params?: { student_id?: string }) { return client.get<TutoringSessionData[]>('/v1/tutoring/sessions', { params }); }
export function createSession(data: Partial<TutoringSessionData>) { return client.post<TutoringSessionData>('/v1/tutoring/sessions', data); }
export function updateSession(id: string, data: Partial<TutoringSessionData>) { return client.patch<TutoringSessionData>(`/v1/tutoring/sessions/${id}`, data); }
export function listPackages(params?: { student_id?: string }) { return client.get<TutoringPackageData[]>('/v1/tutoring/packages', { params }); }
export function createPackage(data: Partial<TutoringPackageData>) { return client.post<TutoringPackageData>('/v1/tutoring/packages', data); }
export function generateInvoice(data: { student_id: string; period_label?: string; notes?: string }) { return client.post<TutoringInvoiceData>('/v1/tutoring/invoices/generate', data); }
export function listInvoices() { return client.get<TutoringInvoiceData[]>('/v1/tutoring/invoices'); }
export function downloadInvoicePdf(invoiceId: string): string { return `/api/v1/tutoring/invoices/${invoiceId}/pdf`; }
